src/librustc/ty/context.rs

```diff
@@ -2694,6 +2695,34 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         self.object_lifetime_defaults_map(id.owner)
             .and_then(|map| map.get(&id.local_id).cloned())
     }
+
+    pub fn va_list_types(&self) -> Vec<Ty<'tcx>> {
```

Comment:
Used to get the `Ty`s of the `VaList` fields. I'm still working on figuring out what to do here for the pointer variants. I'm not sure if the current implementation works for them. My first attempt only used `Type::va_list`, but the `VaList` function argument would get optimized out as a ZST, so I added extra logic here so that the `LayoutDetails` indicate that the `VaList` is not a ZST. Comments and critiques on this would be super helpful.